File: bindings/csharp/Serialization/JsonDeserializer.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections;
using System.Linq;
using System.Reflection;

namespace MikanXR
{
	public class JsonDeserializer
	{
		private class JsonReadVisitor : IVisitor
		{
			private JToken _jsonToken;
			public JToken jsonToken => _jsonToken;

			public JsonReadVisitor(JToken jsonToken)
			{
				_jsonToken = jsonToken;
			}

			public void visitClass(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);
				Type fieldType = accessor.ValueType;
				string fieldName = accessor.ValueName;

				if (jsonToken.Type == JTokenType.Object)
				{
					accessor.ensureValueAllocated();
					object fieldInstance = accessor.getValueObject();

					JsonReadVisitor jsonVisitor = new JsonReadVisitor(jsonToken as JObject);
					Utils.visitObject(fieldInstance, fieldType, jsonVisitor);
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitClass() Field " + fieldName +
						" missing corresponding json object");
				}
			}

			public void visitList(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);
				Type fieldType = accessor.ValueType;
				string fieldName = accessor.ValueName;

				if (jsonToken.Type == JTokenType.Array)
				{
					var jsonArray = jsonToken as JArray;

					accessor.ensureValueAllocated();
					var list= accessor.getValueObject() as IList;
					Type elementType = fieldType.GenericTypeArguments[0];

					list.Clear();
					foreach (var jsonElement in jsonArray)
					{
						// Make a fake "field" for an element in the array
						var elementAccessor = new ValueAccessor(null, elementType);

						// Deserialize the element
						JsonReadVisitor elementVisitor = new JsonReadVisitor(jsonElement);
						Utils.visitValue(elementAccessor, elementVisitor);

						// Add the element to the list
						list.Add(elementAccessor.ValueInstance);
					}
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitList() Field " + fieldName +
						" missing corresponding json array");
				}
			}

			public void visitDictionary(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);
				Type fieldType = accessor.ValueType;
				string fieldName = accessor.ValueName;

				if (jsonToken.Type == JTokenType.Array)
				{
					var jsonMapPairArray = jsonToken as JArray;

					accessor.ensureValueAllocated();
					var map = accessor.getValueObject() as IDictionary;
					Type keyType = fieldType.GenericTypeArguments[0];
					Type valueType = fieldType.GenericTypeArguments[1];

					map.Clear();
					foreach (var jsonPair in jsonMapPairArray)
					{
						// Make a fake "field" for the key
						var keyAccessor = new ValueAccessor(null, keyType);

						// Deserialize the key
						var jsonKey= jsonPair["key"];
						JsonReadVisitor keyVisitor = new JsonReadVisitor(jsonKey);
						Utils.visitValue(keyAccessor, keyVisitor);

						// Make a fake "field" for the value
						var valueAccessor = new ValueAccessor(null, valueType);

						// Deserialize the value
						var jsonValue = jsonPair["value"];
						JsonReadVisitor valueVisitor = new JsonReadVisitor(jsonValue);
						Utils.visitValue(valueAccessor, valueVisitor);

						// Add the key-value pair to the map
						map.Add(keyAccessor.ValueInstance, valueAccessor.ValueInstance);
					}
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitList() Field " + fieldName +
						" missing corresponding json array");
				}
			}

			public void visitPolymorphicObject(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);
				Type fieldType = accessor.ValueType;
				string fieldName = accessor.ValueName;

				accessor.ensureValueAllocated();
				object fieldInstance = accessor.getValueObject();

				// Allocate the element instance using the runtime class name
				Type elementStaticType = fieldType.GenericTypeArguments[0];
				var elementRuntimeTypeName = (string)jsonToken["class_name"];
				var elementRuntimeTypes = 
						from t in elementStaticType.Assembly.GetTypes()
						where 
							t.IsClass && 
							t.Name == elementRuntimeTypeName &&
							t.IsSubclassOf(elementStaticType)
						select t;
				var elementRuntimeType = elementRuntimeTypes.FirstOrDefault();
				if (elementRuntimeType == null)
				{
					throw new Exception(
							"JsonReadVisitor::visitPolymorphicObject() " +
							"ObjectPtr Accessor " + fieldName +
							" used an unknown runtime class_name: " + elementRuntimeTypeName +
							", static class_name: " + elementStaticType.Name);
				}
				var elementInstance = Activator.CreateInstance(elementRuntimeType);

				// Deserialize the element into the newly allocated instance
				JObject elementJson = jsonToken["value"] as JObject;
				var elementVisitor = new JsonReadVisitor(elementJson);
				Utils.visitObject(elementInstance, elementRuntimeType, elementVisitor);

				// Assign the child object to the SerializableObject field
				MethodInfo setInstanceMethod = fieldType.GetMethod("setInstance");
				setInstanceMethod.Invoke(fieldInstance, new object[] { elementInstance });
			}

			public void visitEnum(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);

				if (jsonToken.Type == JTokenType.Integer)
				{
					int value = (int)jsonToken;
					var enumValue= Enum.ToObject(accessor.ValueType, value);

					accessor.setValueObject(enumValue);
				}
				else if (jsonToken.Type == JTokenType.String)
				{
					string value = (string)jsonToken;
					var enumValue= Enum.Parse(accessor.ValueType, value);

					accessor.setValueObject(enumValue);
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitEnum() " +
						"Enum Accessor " + accessor.ValueName +
						" was not a Integer or String json value");
				}
			}

			public void visitBool(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);

				if (jsonToken.Type == JTokenType.Boolean)
				{
					bool value = (bool)jsonToken;

					accessor.setValue(value);
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitBool() " +
						"Bool Accessor " + accessor.ValueName +
						" was not a bool json value");
				}
			}

			public void visitByte(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);

				if (jsonToken.Type == JTokenType.Integer)
				{
					sbyte value = (sbyte)(short)jsonToken;

					accessor.setValue(value);
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitByte() " +
						"Byte Accessor " + accessor.ValueName +
						" was not a Integer json value");
				}
			}

			public void visitUByte(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);

				if (jsonToken.Type == JTokenType.Integer)
				{
					byte value = (byte)(ushort)jsonToken;

					accessor.setValue(value);
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitUByte() " +
						"UByte Accessor " + accessor.ValueName +
						" was not a Integer json value");
				}
			}

			public void visitShort(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);

				if (jsonToken.Type == JTokenType.Integer)
				{
					short value = (short)jsonToken;

					accessor.setValue(value);
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitShort() " +
						"Short Accessor " + accessor.ValueName +
						" was not a Integer json value");
				}
			}

			public void visitUShort(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);

				if (jsonToken.Type == JTokenType.Integer)
				{
					ushort value = (ushort)jsonToken;

					accessor.setValue(value);
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitUShort() " +
						"UShort Accessor " + accessor.ValueName +
						" was not a Integer json value");
				}
			}

			public void visitInt(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);

				if (jsonToken.Type == JTokenType.Integer)
				{
					int value = (int)jsonToken;

					accessor.setValue(value);
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitInt() " +
						"Int Accessor " + accessor.ValueName +
						" was not a Integer json value");
				}
			}

			public void visitUInt(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);

				if (jsonToken.Type == JTokenType.Integer)
				{
					uint value = (uint)jsonToken;

					accessor.setValue(value);
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitUInt() " +
						"UInt Accessor " + accessor.ValueName +
						" was not a Integer json value");
				}
			}

			public void visitLong(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);

				if (jsonToken.Type == JTokenType.Integer)
				{
					long value = (long)jsonToken;

					accessor.setValue(value);
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitLong() " +
						"Long Accessor " + accessor.ValueName +
						" was not a Integer json value");
				}
			}

			public void visitULong(ValueAccessor accessor)
			{
				throw new Exception(
					"JsonReadVisitor::visitULong() " +
					"ULong Accessor " + accessor.ValueName +
					" type not supported by all JSON libraries");
			}

			public void visitFloat(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);

				if (jsonToken.Type == JTokenType.Integer ||
					jsonToken.Type == JTokenType.Float)
				{
					float value = (float)jsonToken;

					accessor.setValue(value);
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitFloat() " +
						"Float Accessor " + accessor.ValueName +
						" was not a Float json value");
				}
			}

			public void visitDouble(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);

				if (jsonToken.Type == JTokenType.Integer ||
					jsonToken.Type == JTokenType.Float)
				{
					double value = (double)jsonToken;

					accessor.setValue(value);
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitDouble() " +
						"Double Accessor " + accessor.ValueName +
						" was not a Double json value");
				}
			}

			public void visitString(ValueAccessor accessor)
			{
				JToken jsonToken = getJsonTokenFromAccessor(accessor);

				if (jsonToken.Type == JTokenType.String)
				{
					string value = (string)jsonToken;

					accessor.setValue(value);
				}
				else
				{
					throw new Exception(
						"JsonReadVisitor::visitString() " +
						"String Accessor " + accessor.ValueName +
						" was not a String json value");
				}
			}

			private JToken getJsonTokenFromAccessor(ValueAccessor accessor)
			{
				var field= accessor.ValueField;
				if (field != null)
				{
					var fieldName = field.Name;
					var jsonObject = _jsonToken as JObject;

					if (jsonObject.TryGetValue(fieldName, out var childJsonObject))
					{
						return childJsonObject;
					}

					throw new Exception("Field "+fieldName+" not found in json");
				}
				else
				{
					// The json object should contain the value we want to deserialize
					return _jsonToken;
				}
			}
		}

		public static bool deserializeFromJsonString<T>(string jsonString, T instance) where T : class
		{
			return deserializeFromJsonString(jsonString, instance, typeof(T));
		}

		public static bool deserializeFromJsonString(
			string jsonString, 
			object instance, 
			Type structType)
		{		
			try
			{
				JObject jsonObject = JObject.Parse(jsonString);
				return deserializeFromJson(jsonObject, instance, structType);
			}
			catch (Exception e)
			{
				Console.WriteLine("Failed to deserialize from JSON string: " + e.Message);
				return false;
			}
		}

		public static bool deserializeFromJson<T>(JObject jsonObject, T instance) where T : class
		{
			return deserializeFromJson(jsonObject, instance, typeof(T));
		}

		public static bool deserializeFromJson(JObject jsonObject, object instance, Type structType)
		{
			try
			{
				var visitor = new JsonReadVisitor(jsonObject);

				Utils.visitObject(instance, structType, visitor);

				return true;
			}
			catch (Exception e)
			{
				Console.WriteLine("Failed to deserialize from JSON: " + e.Message);
				return false;
			}
		}
	}
}

File: bindings/typescript/Serialization/JsonDeserializer.ts
import {
  ValueAccessor,
  IVisitor,
  visitObject,
  visitValue,
  FieldMetadata,
  getSerializationMetadata
} from './SerializationUtils';
import { PolymorphicObject, PolymorphicStruct } from '../PolymorphicObject';

/**
 * Type registry for runtime type lookup
 */
export class TypeRegistry {
  private static types = new Map<string, any>();

  static register(name: string, type: any): void {
    this.types.set(name, type);
  }

  static get(name: string): any | undefined {
    return this.types.get(name);
  }

  static has(name: string): boolean {
    return this.types.has(name);
  }
}

/**
 * JSON read visitor for deserializing objects from JSON
 */
class JsonReadVisitor implements IVisitor {
  private jsonToken: any;

  constructor(jsonToken: any) {
    this.jsonToken = jsonToken;
  }

  getJsonToken(): any {
    return this.jsonToken;
  }

  visitClass(accessor: ValueAccessor): void {
    const jsonToken = this.getJsonTokenFromAccessor(accessor);

    if (typeof jsonToken === 'object' && jsonToken !== null) {
      accessor.ensureValueAllocated();
      const fieldInstance = accessor.getValueObject();

      const jsonVisitor = new JsonReadVisitor(jsonToken);
      visitObject(fieldInstance, accessor.valueType, jsonVisitor);
    } else {
      throw new Error(
        `JsonReadVisitor::visitClass() Field ${accessor.valueName} missing corresponding json object`
      );
    }
  }

  visitList(accessor: ValueAccessor): void {
    const jsonToken = this.getJsonTokenFromAccessor(accessor);

    if (Array.isArray(jsonToken)) {
      accessor.ensureValueAllocated();
      const list = accessor.getValueObject() as any[];

      // Clear the list
      list.length = 0;

      for (const jsonElement of jsonToken) {
        // Determine element type - this is simplified, in real implementation
        // we'd need more type information
        let element: any;

        if (typeof jsonElement === 'object' && jsonElement !== null) {
          // For complex types, we need the metadata
          // For now, create a plain object
          element = jsonElement;
        } else {
          element = jsonElement;
        }

        list.push(element);
      }
    } else {
      throw new Error(
        `JsonReadVisitor::visitList() Field ${accessor.valueName} missing corresponding json array`
      );
    }
  }

  visitDictionary(accessor: ValueAccessor): void {
    const jsonToken = this.getJsonTokenFromAccessor(accessor);

    if (Array.isArray(jsonToken)) {
      accessor.ensureValueAllocated();
      const map = accessor.getValueObject() as Map<any, any>;

      map.clear();

      for (const jsonPair of jsonToken) {
        const key = jsonPair.key;
        const value = jsonPair.value;

        map.set(key, value);
      }
    } else {
      throw new Error(
        `JsonReadVisitor::visitDictionary() Field ${accessor.valueName} missing corresponding json array`
      );
    }
  }

  visitPolymorphicObject(accessor: ValueAccessor): void {
    const jsonToken = this.getJsonTokenFromAccessor(accessor);

    accessor.ensureValueAllocated();
    const fieldInstance = accessor.getValueObject() as PolymorphicObject;

    const elementRuntimeTypeName = jsonToken.class_name || '';
    if (elementRuntimeTypeName === '') {
      return;
    }

    // Look up the runtime type
    const elementRuntimeType = TypeRegistry.get(elementRuntimeTypeName);
    if (!elementRuntimeType) {
      throw new Error(
        `JsonReadVisitor::visitPolymorphicObject() ` +
        `PolymorphicObject Accessor ${accessor.valueName} ` +
        `used an unknown runtime class_name: ${elementRuntimeTypeName}`
      );
    }

    // Create instance
    const elementInstance = new elementRuntimeType();

    // Deserialize the element
    const elementJson = jsonToken.value;
    const elementVisitor = new JsonReadVisitor(elementJson);
    visitObject(elementInstance, elementRuntimeType, elementVisitor);

    // Set the instance on the PolymorphicObject
    const classId = BigInt(jsonToken.class_id || '0');
    fieldInstance.setInstance(elementInstance, classId);
  }

  visitEnum(accessor: ValueAccessor): void {
    const jsonToken = this.getJsonTokenFromAccessor(accessor);

    if (typeof jsonToken === 'number') {
      accessor.setValueObject(jsonToken);
    } else if (typeof jsonToken === 'string') {
      // Parse string enum value
      // In TypeScript, we'd need the enum object to convert
      accessor.setValueObject(jsonToken);
    } else {
      throw new Error(
        `JsonReadVisitor::visitEnum() Enum Accessor ${accessor.valueName} ` +
        `was not a number or string json value`
      );
    }
  }

  visitBool(accessor: ValueAccessor): void {
    const jsonToken = this.getJsonTokenFromAccessor(accessor);

    if (typeof jsonToken === 'boolean') {
      accessor.setValue(jsonToken);
    } else {
      throw new Error(
        `JsonReadVisitor::visitBool() Bool Accessor ${accessor.valueName} ` +
        `was not a boolean json value`
      );
    }
  }

  visitByte(accessor: ValueAccessor): void {
    this.visitNumeric(accessor, 'Byte');
  }

  visitUByte(accessor: ValueAccessor): void {
    this.visitNumeric(accessor, 'UByte');
  }

  visitShort(accessor: ValueAccessor): void {
    this.visitNumeric(accessor, 'Short');
  }

  visitUShort(accessor: ValueAccessor): void {
    this.visitNumeric(accessor, 'UShort');
  }

  visitInt(accessor: ValueAccessor): void {
    this.visitNumeric(accessor, 'Int');
  }

  visitUInt(accessor: ValueAccessor): void {
    this.visitNumeric(accessor, 'UInt');
  }

  visitLong(accessor: ValueAccessor): void {
    const jsonToken = this.getJsonTokenFromAccessor(accessor);

    if (typeof jsonToken === 'number' || typeof jsonToken === 'string') {
      const value = BigInt(jsonToken);
      accessor.setValue(value);
    } else {
      throw new Error(
        `JsonReadVisitor::visitLong() Long Accessor ${accessor.valueName} ` +
        `was not a number or string json value`
      );
    }
  }

  visitULong(accessor: ValueAccessor): void {
    const jsonToken = this.getJsonTokenFromAccessor(accessor);

    if (typeof jsonToken === 'number' || typeof jsonToken === 'string') {
      const value = BigInt(jsonToken);
      accessor.setValue(value);
    } else {
      throw new Error(
        `JsonReadVisitor::visitULong() ULong Accessor ${accessor.valueName} ` +
        `was not a number or string json value`
      );
    }
  }

  visitFloat(accessor: ValueAccessor): void {
    this.visitNumeric(accessor, 'Float');
  }

  visitDouble(accessor: ValueAccessor): void {
    this.visitNumeric(accessor, 'Double');
  }

  visitString(accessor: ValueAccessor): void {
    const jsonToken = this.getJsonTokenFromAccessor(accessor);

    if (typeof jsonToken === 'string') {
      accessor.setValue(jsonToken);
    } else {
      throw new Error(
        `JsonReadVisitor::visitString() String Accessor ${accessor.valueName} ` +
        `was not a string json value`
      );
    }
  }

  private visitNumeric(accessor: ValueAccessor, typeName: string): void {
    const jsonToken = this.getJsonTokenFromAccessor(accessor);

    if (typeof jsonToken === 'number') {
      accessor.setValue(jsonToken);
    } else {
      throw new Error(
        `JsonReadVisitor::visit${typeName}() ${typeName} Accessor ${accessor.valueName} ` +
        `was not a number json value`
      );
    }
  }

  private getJsonTokenFromAccessor(accessor: ValueAccessor): any {
    const field = accessor.valueField;
    if (field !== null) {
      if (this.jsonToken.hasOwnProperty(field)) {
        return this.jsonToken[field];
      }
      throw new Error(`Field ${field} not found in json`);
    } else {
      return this.jsonToken;
    }
  }
}

/**
 * Deserialize an object from JSON string
 */
export function deserializeFromJsonString(
  jsonString: string,
  instance: any,
  instanceType?: any
): boolean {
  try {
    const jsonObject = JSON.parse(jsonString);
    return deserializeFromJson(jsonObject, instance, instanceType);
  } catch (e) {
    console.error('Failed to deserialize from JSON string:', e);
    return false;
  }
}

/**
 * Deserialize an object from JSON object
 */
export function deserializeFromJson(
  jsonObject: any,
  instance: any,
  instanceType?: any
): boolean {
  try {
    const type = instanceType || instance?.constructor;
    const visitor = new JsonReadVisitor(jsonObject);

    visitObject(instance, type, visitor);

    return true;
  } catch (e) {
    console.error('Failed to deserialize from JSON:', e);
    return false;
  }
}
